### DISCUSSION TITLE
Using Expander Emulation and triggering zones

### DISCUSSION THREAD

**Discussion by abishur** [Dilbert66/esphome-dsckeybus#55] [Q&A]
I noticed in the new branch you mentioned zone expander emulation and wanted to make sure I understood it correctly:

By adding these two/uncommenting these two section in the yaml file
` #zone expander addresses:
  # 9  - zones 9-16
  # 10 - zones 17-24
  # 11 - zones 25-32
  # 12 - zones 33-40 (for systems with 64 zone support)
  # 13 - zones 41-48 (for systems with 64 zone support)
  # 14 - zones 49-56 (for systems with 64 zone support)
  # 16 - zones 57-64 (for systems with 64 zone support)  
  expanderAddr1: "10" # 1st zone expander emulator address to use . Set to 0 to disable.

 - platform: gpio  
    pin: D8
    id: pind8
    device_class: window
    on_press:       #pin high=on(open), pin low=off(closed)
      - lambda: |-
          dsc.setZoneFault(17,1);  #set zone 17 as open
    on_release:
      - lambda: |-
          dsc.setZoneFault(17,0);  #set zone 17 as closed

`
Then the input to pin D8 in this example would "trigger" zone 17 where a high value means that the device is open.

First off that's EXTREMELY cool, second is there a way that I can combine this with some wireless zigbee sensors?  It looks like I could using the MQTT payload:

`{"zone":<zone#>,"fault":<on:off or 0:1>}`

But in that scenario I would need to do something other than use - platform: gpio, right?

**Comment by abishur** [comment]
I think I realized the answer to half of my question, to create a virtual zone that isn't attached to a physical input, I just use the template platform the way that's already been done for the actual zones like this:

`- platform: template
    id: z10
    device_class: door
    name: "$systemName Backdoor Open"
  `

And then I use the MQTT in Home Assistant to set it high and low based on the zigbee sensor, am I on the right track here?

**Reply by Dilbert66** [reply]
You need to create your  zone on the ESP so that you can have HA see that zones status, just like any other zones but using the extended zone that you want to use.

To trigger a fault using MQTT, call the HA MQTT service with this topic:  "<yoursystemnameinesphome>/alarm/set"  and with this payload "{"zone":<zone#>,"fault":<on:off or 0:1>}" where <zone#> is your virtual zone .  If you use an expander address of 10, then your zone range is 17-24.

I would recommend you use the API interface instead of MQTT if you are using HA.   The MQTT support is really for non HA environments.  I havent done much testing with it.    To trigger a fault using the API, you use the set_zone_fault service.


The gpio example is only there to show that it is also possible to extend the alarm system using the ESP extra pins as trigger points.

**Reply by abishur** [reply]
Oh, duh.  I'm not sure why I was being so brain dead on that.  Thanks!

**Comment by abishur** [comment]
Okay I've enabled the zone expander using

`expanderAddr1: "9" # 1st zone expander emulator address to use . Set to 0 to disable. `

Then I added two new zones, one will be a sensor for my backdoor triggered by a zigbee device from Home Assistant and the other will be a GPIO input from my smoke alarms.  Using the following code:

```
  - platform: template
    id: z9
    device_class: door
    name: "Backdoor Open(z9)"

  - platform: gpio  
    pin: GPIO27
    id: f1
    device_class: smoke
    name: "Smoke Alarm Active(z10)"
#pin high=on(smoke detected), pin low=off(normal)
#set zone 10 as on    
    on_press:       
      - lambda: |-
          dsc.setZoneFault(10,0);
#set zone 10 as off    
    on_release:
      - lambda: |-
          dsc.setZoneFault(10,1);
```

The smoke alarm one shows its status, though I think I might have the values backwards, but z9 has it's state reported as "unknown" in HA.  Any ideas why Zone 9 is having issues?

**Reply by abishur** [reply]
@Dilbert66 I hope I'm not out of line to ping you, I realized that by commenting in a thread that I had left off as solved my follow up question may have slipped under the radar (no rush if you had seen it and haven't had time to look into it).  Any thoughts on why z9 in the above code snippet would exist as an "unknown state" in home assistant?

That is to say that Home assistant correctly lists the entity, but it's state is stuck as "unknown"

**Reply by Dilbert66** [reply]
Sorry, not a clue.  Something in your config is not correct I assume.  Is z9 configured in the onzonestatuschange case statement further up in the config?

**Reply by Dilbert66** [reply]
I assume you've also configured the expander address in your alarm panel config.  If not, see your panel programming panel for that info.

**Reply by abishur** [reply]
> Sorry, not a clue. Something in your config is not correct I assume. Is z9 configured in the onzonestatuschange case statement further up in the config?

That was it!  Once I added zone 9 (and made sure to do that for any other custom zone I feel like creating) it worked!  Thank you!
